# Waldemar Chang - Assignment 8: Integrating the Generation and Retrieval Service
## EN.705.603.82.FA24 Creating AI-Enabled Systems
### Task 1: Search nearest neighbors
#### In a notebook called `notebook/context_answering_analysis.ipynb`, demonstrate the output of search_neighbors() function.

In [1]:
import os
import sys
import json
import numpy as np
from pipeline import Pipeline
from IPython.display import display, Markdown

C:\Users\walde\anaconda3\Lib\site-packages\sentence_transformers\cross_encoder\CrossEncoder.py:13: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange


In [2]:
# Define the corpus directory
corpus_directory = r'C:\Users\walde\systems\textwave\storage\corpus'

# Initialize Pipeline
pipeline = Pipeline()
pipeline.preprocess_corpus(corpus_directory)

Generated 146 chunks from 'S08_set3_a1.txt.clean':
Chunk 1: 'John_Adams John Adams, Jr. (October 30,1735 July 4, 1826) was the second President of the United States (1797 1801). He also served as America's first Vice President (1789 1797).'
Chunk 2: 'He was defeated for re-election in the "Revolution of 1800" by Thomas Jefferson. Adams was also the first President to reside in the newly built White House in Washington, D.C., which was completed in 1800.'
Chunk 3: 'Adams, a sponsor of the American Revolution in Massachusetts, was a driving force for independence in 1776; Jefferson called him the "Colossus of Independence". He represented the Continental Congress in Europe.'
Chunk 4: 'He was a major negotiator of the eventual peace treaty with Great Britain, and chiefly responsible for obtaining the loans from the Amsterdam money market necessary for the conduct of the Revolution. His prestige secured his two elections as Washington's Vice President and his election to succeed him.'
Chun

In [3]:
# Function to run search_neighbors with given query and k, then print results
def run_search_analysis(query, k):
    print(f"\nQuery: '{query}', k: {k}")
    query_embedding = pipeline._Pipeline__encode(query)
    neighbors = pipeline.search_neighbors(query_embedding, k)
    for i, neighbor in enumerate(neighbors):
        print(f"Neighbor {i + 1}: {neighbor}")

In [4]:
# Search neighbors for the given queries
queries = [
    ("Who was Abraham Lincoln?", 15),
    ("Who was Abraham Adams?", 15),
    ("Did Abraham Lincoln live in the Frontier?", 1),
    ("Did Abraham Lincoln live in the Frontier?", 10),
    ("Did Abraham Lincoln live in the Frontier?", 20),
    ("Did Abraham Lincoln live in the Frontier?", 50),
    ("How did Fillmore ascend to the presidency?", 15),
    ("What is the capital of France?", 15),
]

In [5]:
query, k = queries[0]

# Title for this particular query output
display(Markdown(f"### Running search analysis for query: '{query}' with k={k}"))

# Run the analysis function for the given query and k
run_search_analysis(query, k)

### Running search analysis for query: 'Who was Abraham Lincoln?' with k=15


Query: 'Who was Abraham Lincoln?', k: 15
Found 15 neighbors.
Neighbor 1: Abraham Lincoln Abraham Lincoln (February 12, 1809 – April 15, 1865) was the sixteenth President of the United States, serving from March 4, 1861 until his assassination. As an outspoken opponent of the expansion of slavery in the United States, "[I]n his short autobiography written for the 1860 presidential campaign, Lincoln would describe his protest in the Illinois legislature as one that 'briefly defined his position on the slavery question, and so far as it goes, it was then the same that it is now."
Neighbor 2: Lincoln is well known for ending slavery in the United States. In 1861 – 1862, however, he made it clear that the North was fighting the war to preserve the Union, not to abolish slavery.
Neighbor 3: Lincoln, Illinois, is the only city to be named for Abraham Lincoln before he became President. Lincoln's name and image appear in numerous places.
Neighbor 4: Lincoln won the Republican Party nomination

### Query 1 Analysis
The search analysis for the query "Who was Abraham Lincoln?" with k=15 yielded 15 neighbors that collectively provide insights into Abraham Lincoln's life, career, and impact. The output effectively highlighted several key aspects, such as Lincoln's role as the sixteenth President of the United States, his leadership during the Civil War, and his efforts to end slavery in the country.

The search results succeeded in providing a foundational overview of Lincoln's biography, offering information on both his political achievements and his personal background. This included details on his humble beginnings, his rise to prominence, and notable actions taken during his presidency. Additionally, the results captured Lincoln’s legacy, touching on his assassination, his portrayal as a martyr for national unity, and his lasting influence on American history.

However, the analysis of these neighbors also reveals some limitations in the search approach. One issue is the repetition of core details about Lincoln's presidency and political achievements across multiple neighbors, which reduces the diversity and uniqueness of the information provided. Another challenge is the inclusion of tangential information, such as references to Lincoln, Illinois, which, while related to Lincoln's legacy, may be less relevant to the central query.

In [6]:
query, k = queries[1]

# Title for this particular query output
display(Markdown(f"### Running search analysis for query: '{query}' with k={k}"))

# Run the analysis function for the given query and k
run_search_analysis(query, k)

### Running search analysis for query: 'Who was Abraham Adams?' with k=15


Query: 'Who was Abraham Adams?', k: 15
Found 15 neighbors.
Neighbor 1: Adams, a sponsor of the American Revolution in Massachusetts, was a driving force for independence in 1776; Jefferson called him the "Colossus of Independence". He represented the Continental Congress in Europe.
Neighbor 2: His birthplace is now part of Adams National Historical Park. His father, a farmer and a Deacon, also named John (1690-1761), was a fourth-generation descendant of Henry Adams, who immigrated from Barton St David, Somerset, England, to Massachusetts Bay Colony in about 1636, from a Welsh male line called Ap Adam.
Neighbor 3: The Works of John Adams (1854), vol III, p 421, diary entry for July 26, 1796. The Unitarian Universalist Historical Society sheds some light on Adams’s religious beliefs.
Neighbor 4: Their children were Abigail (1765-1813); future president John Quincy (1767-1848); Susanna (1768–1770); Charles (1770-1800); Thomas Boylston (1772-1832); and Elizabeth (1775) who was stillborn.

### Query 2 Analysis
The search analysis for the query "Who was Abraham Adams?" with k=15 revealed 15 neighbors that provided insights mainly about the historical figure John Adams. The majority of the neighbors focused on his role in the American Revolution, his political career, and his personal life, including his family relationships and his stance on major social issues of the time.

The retrieved neighbors offered a broad overview of Adams's influence during the American Revolution, highlighting his contributions as a "driving force for independence" and a key representative in the Continental Congress. These details emphasize his prominent role in American history, including his significant influence in drafting key founding documents such as the Declaration of Independence. This level of information effectively answers a query about his historical importance and achievements.

The results also shed light on John Adams's personality and personal experiences. For instance, his contentious nature and political career struggles are highlighted, which provides a deeper look into the challenges he faced during his time in public office. The retrieved neighbors also provide information on his reconciliation with Thomas Jefferson, illustrating a significant episode in his later life. Additionally, the neighbors explore Adams's family, including his children, and his wife Abigail Adams’s views on slavery, which adds richness to his historical portrait.

However, there are certain limitations to the retrieved information. The primary issue is that the query asks about "Abraham Adams," but the system retrieved information predominantly about John Adams. This suggests that the system's ability to discern between similar names or correctly resolve less common queries could be improved. Moreover, the neighbors contain some redundancy, as several of the entries repeat similar information, such as Adams's contributions to independence, his family life, and his political career. This repetition reduces the diversity of content and limits the breadth of information provided.

In [7]:
query, k = queries[2]

# Title for this particular query output
display(Markdown(f"### Running search analysis for query: '{query}' with k={k}"))

# Run the analysis function for the given query and k
run_search_analysis(query, k)

### Running search analysis for query: 'Did Abraham Lincoln live in the Frontier?' with k=1


Query: 'Did Abraham Lincoln live in the Frontier?', k: 1
Found 1 neighbors.
Neighbor 1: This area was at the time considered the "frontier." The name "Abraham" was chosen to commemorate his grandfather, who was killed in an American Indian raid in 1786.
Neighbor 1: This area was at the time considered the "frontier." The name "Abraham" was chosen to commemorate his grandfather, who was killed in an American Indian raid in 1786.


### Query 3 Analysis

The search analysis for the query "Did Abraham Lincoln live in the Frontier?" with k=1 revealed a single neighbor that provided direct and relevant information about Lincoln's connection to the frontier. The retrieved neighbor specifically mentions that the area in question was considered the "frontier" and provides additional context by explaining that the name "Abraham" was chosen to honor Lincoln's grandfather, who was killed in an Indian raid in 1786.

This result offers a concise and relevant response to the query, effectively establishing Lincoln's association with the frontier, which is the main focus of the question. The mention of his grandfather’s death also adds a meaningful historical detail that enriches the narrative. Given that the query was narrowly focused and k=1 was used, the system successfully returned an answer that was both direct and pertinent.

However, there are notable limitations to the retrieved result. The use of a single neighbor (k=1) restricts the response's depth and comprehensiveness. While the result establishes the connection between Lincoln and the frontier, it lacks elaboration on Lincoln's experiences, daily life, and challenges while living in the frontier. Such additional context would have provided a more complete answer, helping the user understand Lincoln's formative years in greater detail. Increasing k could have allowed the system to retrieve additional information that would paint a fuller picture of Lincoln's life on the frontier.

Another aspect to consider is the similarity metric used in determining the nearest neighbor. Since the retrieval system employed the Euclidean metric, it may have focused primarily on closeness in terms of numerical vector values rather than semantic diversity. This may have contributed to the narrowness of the retrieved content. A different similarity metric, such as cosine similarity, could potentially retrieve results with richer contextual diversity, thereby providing a more well-rounded answer.

In [8]:
query, k = queries[3]

# Title for this particular query output
display(Markdown(f"### Running search analysis for query: '{query}' with k={k}"))

# Run the analysis function for the given query and k
run_search_analysis(query, k)

### Running search analysis for query: 'Did Abraham Lincoln live in the Frontier?' with k=10


Query: 'Did Abraham Lincoln live in the Frontier?', k: 10
Found 10 neighbors.
Neighbor 1: This area was at the time considered the "frontier." The name "Abraham" was chosen to commemorate his grandfather, who was killed in an American Indian raid in 1786.
Neighbor 2: He later noted that this move was "partly on account of slavery," and partly because of difficulties with land deeds in Kentucky: Unlike land in the Northwest Territory, Kentucky never had a proper U.S. survey, and farmers often had difficulties proving title to their property. Lincoln was only nine when his mother, then thirty-four years old, died of milk sickness.
Neighbor 3: Abraham Lincoln Online Retrieved on March 11, 2007 For several months, Lincoln ran a small store in New Salem, selling tea, coffee, sugar, salt, blue calico, brown muslin, straw hats and whiskey. Beveridge (1928) 1:127-8 Later, he found work as village postmaster and as a surveyor.
Neighbor 4: Lincoln believed in the Whig theory of the presidency, 

### Query 4 Analysis

The search analysis for the query "Did Abraham Lincoln live in the Frontier?" with k=10 revealed ten neighbors, providing both direct information about Lincoln's frontier life and broader historical context.

Neighbor 1 directly establishes that Lincoln lived in what was then considered the "frontier" and provides background on his grandfather's death. Neighbor 2 adds that Lincoln's family moved from Kentucky due to land issues and provides a personal detail about his mother's death. Neighbor 3 describes Lincoln's roles in New Salem, including his work as a storekeeper, postmaster, and surveyor, illustrating his integration into the frontier community.

Neighbor 4, while more political, mentions Lincoln's signing of the Homestead Act, relevant to his involvement in westward expansion. Neighbor 5 gives insight into Lincoln's family's anti-slavery sentiment, adding context about the values that influenced his early life. Neighbor 6 discusses Lincoln's humble frontier beginnings, which helped build his "Honest Abe" persona.

The remaining neighbors are less directly relevant, including information on Lincoln's plans for colonization (Neighbor 7), his funeral (Neighbor 8), and political analysis (Neighbors 9 and 10). While significant, these details do not contribute directly to the question of Lincoln's frontier experience.

Overall, the retrieved neighbors provide a mix of directly relevant information about Lincoln's frontier life and unrelated details, suggesting the need for more focused search results.

In [9]:
query, k = queries[4]

# Title for this particular query output
display(Markdown(f"### Running search analysis for query: '{query}' with k={k}"))

# Run the analysis function for the given query and k
run_search_analysis(query, k)

### Running search analysis for query: 'Did Abraham Lincoln live in the Frontier?' with k=20


Query: 'Did Abraham Lincoln live in the Frontier?', k: 20
Found 20 neighbors.
Neighbor 1: This area was at the time considered the "frontier." The name "Abraham" was chosen to commemorate his grandfather, who was killed in an American Indian raid in 1786.
Neighbor 2: He later noted that this move was "partly on account of slavery," and partly because of difficulties with land deeds in Kentucky: Unlike land in the Northwest Territory, Kentucky never had a proper U.S. survey, and farmers often had difficulties proving title to their property. Lincoln was only nine when his mother, then thirty-four years old, died of milk sickness.
Neighbor 3: Abraham Lincoln Online Retrieved on March 11, 2007 For several months, Lincoln ran a small store in New Salem, selling tea, coffee, sugar, salt, blue calico, brown muslin, straw hats and whiskey. Beveridge (1928) 1:127-8 Later, he found work as village postmaster and as a surveyor.
Neighbor 4: Lincoln believed in the Whig theory of the presidency, 

### Query 5 Analysis

The search analysis for the query "Did Abraham Lincoln live in the Frontier?" with k=20 produced a set of neighbors, many of which overlapped with those found in previous queries, particularly Query 4 (k=10). This overlap provides an opportunity to compare the relevance of the results across different k values.

Neighbor 1 is again the most directly relevant, providing clear information about Lincoln's frontier experience and mentioning the naming of "Abraham" after his grandfather, who was killed in an American Indian raid. Neighbor 2 repeats from Query 4, discussing the family's move from Kentucky and the land deed issues. Neighbor 3 describes Lincoln's roles in New Salem, reiterating his life in a frontier community. Neighbor 6 emphasizes Lincoln's "pioneer background," which also appeared in Query 4.

Neighbor 17 describes Lincoln's family's move to a new homestead in Illinois and his journey to New Salem, adding depth to the understanding of Lincoln's frontier experience. Neighbor 13 provides additional context on Lincoln's early life in Indiana, which was also frontier-like.

The expanded k=20 search brought in new neighbors that provide mixed relevance. Neighbor 11 mentions Lincoln's stepmother, Sarah Bush Johnston, adding personal context to Lincoln's upbringing in a frontier environment. Neighbor 12 touches on Lincoln's moral beliefs, indirectly connecting to values that may have influenced his frontier life. Neighbor 15 discusses Lincoln's father's pro-slavery stance, contrasting with Lincoln's beliefs and adding background on the challenges Lincoln faced in shaping his views while living in a frontier setting.

Neighbor 17 describes the difficult winter the Lincoln family endured before moving to Illinois and Lincoln's journey to New Salem, reinforcing the idea of constant movement and struggle. Neighbor 19 highlights Lincoln's aversion to hunting and fishing, suggesting a nuanced aspect of his frontier life. Neighbor 18 provides insight into Lincoln's intellectual pursuits, emphasizing his dedication to self-education despite the hardships of frontier life.

However, many additional neighbors are less directly relevant. For example, Neighbor 20 describes Lincoln being celebrated by freed slaves, which does not contribute to understanding his frontier experience. Neighbor 16 discusses Lincoln's political stance on economic advancement and territorial expansion, which, while related to frontier expansion, is not specific to his personal frontier life.

The comparison between Queries 4 and 5 demonstrates that increasing the value of k results in a wider array of information but at the cost of more tangential content. Query 5 included many of the same relevant pieces of information found in Query 4, but with additional, less focused content that diluted the overall relevance. For a narrow query like Lincoln's frontier experience, a smaller k value might provide more concentrated and useful results.

In [10]:
query, k = queries[5]

# Title for this particular query output
display(Markdown(f"### Running search analysis for query: '{query}' with k={k}"))

# Run the analysis function for the given query and k
run_search_analysis(query, k)

### Running search analysis for query: 'Did Abraham Lincoln live in the Frontier?' with k=50


Query: 'Did Abraham Lincoln live in the Frontier?', k: 50
Found 50 neighbors.
Neighbor 1: This area was at the time considered the "frontier." The name "Abraham" was chosen to commemorate his grandfather, who was killed in an American Indian raid in 1786.
Neighbor 2: He later noted that this move was "partly on account of slavery," and partly because of difficulties with land deeds in Kentucky: Unlike land in the Northwest Territory, Kentucky never had a proper U.S. survey, and farmers often had difficulties proving title to their property. Lincoln was only nine when his mother, then thirty-four years old, died of milk sickness.
Neighbor 3: Abraham Lincoln Online Retrieved on March 11, 2007 For several months, Lincoln ran a small store in New Salem, selling tea, coffee, sugar, salt, blue calico, brown muslin, straw hats and whiskey. Beveridge (1928) 1:127-8 Later, he found work as village postmaster and as a surveyor.
Neighbor 4: Lincoln believed in the Whig theory of the presidency, 

### Query 6 Analysis

The search analysis for the query "Did Abraham Lincoln live in the Frontier?" with k=50 produced a diverse set of neighbors, many of which overlap with previous queries, particularly Queries 4 and 5. However, expanding the value of k to 50 introduced additional perspectives, including more nuanced details about Lincoln's life and broader historical context, albeit with more tangential information.

In addition to the previously discussed neighbors, the expanded k=50 search includes new details that add depth to Lincoln's frontier experience. Neighbor 27 mentions the Sinking Spring Farm, which Lincoln's father purchased, now preserved as part of the Abraham Lincoln Birthplace National Historic Site. This adds another layer of context to Lincoln's early life on the frontier.

Neighbor 28 discusses the family's relocation to Macon County, Illinois, due to land-title difficulties, which connects directly to the struggles often faced by frontier settlers. This neighbor adds depth to understanding how these hardships influenced Lincoln's decision to study law. Neighbor 42 mentions Lincoln's election as captain of an Illinois militia company during the Black Hawk War, further illustrating his involvement in frontier life and community leadership.

The expanded set also introduces less directly relevant neighbors. For example, Neighbor 22 discusses Lincoln's misunderstanding of Southern nationalism, and Neighbor 20 describes Lincoln being celebrated by freed slaves—details that, while significant historically, are not directly tied to his frontier life. Neighbor 35 mentions Lincoln's visit to Richmond after the Civil War, which is unrelated to the specific query about his frontier experience.

Overall, increasing k to 50 provided a broader array of information, including both additional relevant details and a significant number of tangential entries. The new neighbors added depth to Lincoln's frontier narrative, such as the family's repeated relocations and Lincoln's role in the militia. However, many neighbors diverged from the core topic, diluting the focus. This suggests that while a larger k value can enrich the context, it also risks including more irrelevant information, which may reduce the overall utility for specific, narrow queries.

In [11]:
query, k = queries[6]

# Title for this particular query output
display(Markdown(f"### Running search analysis for query: '{query}' with k={k}"))

# Run the analysis function for the given query and k
run_search_analysis(query, k)

### Running search analysis for query: 'How did Fillmore ascend to the presidency?' with k=15


Query: 'How did Fillmore ascend to the presidency?', k: 15
Found 15 neighbors.
Neighbor 1: Instead, Fillmore joined the anti-immigrant, anti-Catholic American Party, the political organ of the Know-Nothing movement. He ran in the election of 1856 as the party's candidate, attempting to win a non-consecutive second term as President (a feat accomplished only once in American politics, by Grover Cleveland).
Neighbor 2: He tried to unite the party by pointing out the differences between the Whigs and the Democrats (by proposing tariff reforms that negatively reflected on the Democratic Party). Another primary objective of Fillmore was to preserve the Union from the intensifying slavery debate.
Neighbor 3: Fillmore signaled this shift by appointing Daniel Webster as his Secretary of State. As president, Fillmore dealt with increasing party divisions within the Whig party; party harmony became one of his primary objectives.
Neighbor 4: Official White House portrait of Millard FillmoreFillm

### Query 7 Analysis

The search analysis for the query "How did Fillmore ascend to the presidency?" with k=15 produced relevant neighbors, detailing Millard Fillmore's path to the presidency and the context of his political career. The most pertinent neighbor for answering the query is Neighbor 4, which clearly states that Fillmore ascended to the presidency upon the sudden death of President Zachary Taylor in July 1850. It also describes the subsequent political shift that occurred when Fillmore replaced Taylor's cabinet to align with his agenda, particularly regarding the Compromise of 1850.

Neighbor 7 further supports this explanation by emphasizing that Fillmore was the second Vice President to assume the presidency due to the death of a sitting president, highlighting the specific circumstances of Taylor's death. Neighbor 13 also mentions Fillmore's role in presiding over the Senate during intense debates leading up to the Compromise of 1850, which ties into the transition in leadership.

Other neighbors provide additional context about Fillmore's career and political maneuvering, which is useful for understanding the political atmosphere during his rise. For example, Neighbor 15 explains his nomination for vice president during the Whig convention in 1848, intended to balance the ticket and heal divisions within the party. This nomination was crucial, as it positioned him to become vice president and later ascend to the presidency. Neighbor 3 adds that Fillmore faced significant party divisions as president and sought to maintain Whig unity.

Several neighbors, while interesting, provide less directly relevant information. Neighbor 1 discusses Fillmore’s post-presidency involvement with the Know-Nothing Party, while Neighbor 5 elaborates on his unsuccessful attempts to gain the presidential nomination after Taylor's term ended. These details, while part of Fillmore's overall political story, are not central to explaining how he initially ascended to the presidency. Similarly, Neighbor 6 touches on Fillmore’s policy initiatives during his presidency, such as enforcing the fugitive slave law, which is more about his presidential actions rather than his rise to power.

Overall, the neighbors provided a clear picture of how Fillmore became president, emphasizing the unexpected death of Zachary Taylor as the catalyst for his ascension. The additional details on his political maneuvering, cabinet appointments, and attempts to maintain party harmony offer useful context but do not directly explain his path to the presidency.

In [12]:
query, k = queries[7]

# Title for this particular query output
display(Markdown(f"### Running search analysis for query: '{query}' with k={k}"))

# Run the analysis function for the given query and k
run_search_analysis(query, k)

### Running search analysis for query: 'What is the capital of France?' with k=15


Query: 'What is the capital of France?', k: 15
Found 15 neighbors.
Neighbor 1: After his term in the Senate, Monroe was appointed Minister to France from 1794 to 1796. His appointment there was made difficult as he had strong sympathies for the French Revolution, but dutifully maintained President Washington's strict policy of neutrality between Britiain and France.
Neighbor 2: In the late stages of the war he took personal control of negotiations with Germany, especially with the Fourteen Points and the Armistice. He went to Paris in 1919 to create the League of Nations and shape the Treaty of Versailles, with special attention on creating new nations out of defunct empires.
Neighbor 3: It also made several suggestions for specific disputes in Europe on the recommendation of Wilson's foreign policy advisor, Colonel Edward M. House, and his team of 150 advisors known as “The Inquiry.” The points were: Abolition of secret treaties Freedom of the seas Free Trade Disarmament Adjustment o

### Query 8 Analysis

The search analysis for the query "What is the capital of France?" with k=15 produced a set of neighbors that were largely unrelated to the direct answer to the query. Instead of providing a simple and factual response ("Paris"), the retrieved results were focused on historical context involving France and its international relations.

Neighbor 1 mentions James Monroe’s appointment as Minister to France, and Neighbor 2 discusses President Wilson’s role at the Treaty of Versailles in Paris. These entries provide some connection to France but fail to directly address the query. Neighbor 3 and Neighbor 8 further elaborate on Wilson's involvement in international diplomacy, including France's role in post-World War I negotiations, which, again, are historically informative but not directly relevant.

Neighbor 4 touches on the historical French support during the American Revolution, while Neighbor 13 describes the undeclared naval Quasi-War between the U.S. and France. These entries are entirely unrelated to the geographical question posed. Neighbors like Neighbor 7, which describes the state nickname for Illinois, are even more off-topic.

Overall, the retrieved information for Query 8 lacks direct relevance, as none of the neighbors provide the simple answer: "Paris." Instead, the responses revolve around broader historical topics involving France. This suggests a potential gap in the corpus or a need for refining the search model to better recognize and retrieve concise, factual answers for direct questions.

### Task 2: Generate answers
#### In a notebook called `notebook/context_answering_analysis.ipynb`, demonstrate the output of generate_answer() with the following.

In [13]:
# Define queries with k values and rerank options
queries = [
    ("Who was Abraham Lincoln?", 15, True),
    ("Who was Abraham Lincoln?", 15, False),
    ("Who was Abraham Adams?", 15, True),
    ("Who was Abraham Adams?", 15, False),
    ("How did Fillmore ascend to the presidency?", 1, True),
    ("How did Fillmore ascend to the presidency?", 5, False),
    ("How did Fillmore ascend to the presidency?", 10, True),
    ("How did Fillmore ascend to the presidency?", 20, False),
    ("What trail did Lincoln use a Farmers' Almanac in?", 1, True),
    ("What trail did Lincoln use a Farmers' Almanac in?", 5, False),
    ("What trail did Lincoln use a Farmers' Almanac in?", 10, True),
    ("What trail did Lincoln use a Farmers' Almanac in?", 20, False),
    ("What is the capital of France?", 15, True),
    ("What is the capital of France?", 15, False)
]

# Run each query and display the results
for query, k, rerank in queries:
    
    # Title for this particular query output
    display(Markdown(f"### Running search analysis for query: '{query}' with k={k}, rerank={rerank}"))
    
    output = pipeline.query(query=query, k=k, rerank=rerank)
    print(output)
    print("-" * 40)

### Running search analysis for query: 'Who was Abraham Lincoln?' with k=15, rerank=True

Found 15 neighbors.
Neighbor 1: Abraham Lincoln Abraham Lincoln (February 12, 1809 – April 15, 1865) was the sixteenth President of the United States, serving from March 4, 1861 until his assassination. As an outspoken opponent of the expansion of slavery in the United States, "[I]n his short autobiography written for the 1860 presidential campaign, Lincoln would describe his protest in the Illinois legislature as one that 'briefly defined his position on the slavery question, and so far as it goes, it was then the same that it is now."
Neighbor 2: Lincoln is well known for ending slavery in the United States. In 1861 – 1862, however, he made it clear that the North was fighting the war to preserve the Union, not to abolish slavery.
Neighbor 3: Lincoln, Illinois, is the only city to be named for Abraham Lincoln before he became President. Lincoln's name and image appear in numerous places.
Neighbor 4: Lincoln won the Republican Party nomination in 1860 and was elected president later t

### Running search analysis for query: 'Who was Abraham Lincoln?' with k=15, rerank=False

Found 15 neighbors.
Neighbor 1: Abraham Lincoln Abraham Lincoln (February 12, 1809 – April 15, 1865) was the sixteenth President of the United States, serving from March 4, 1861 until his assassination. As an outspoken opponent of the expansion of slavery in the United States, "[I]n his short autobiography written for the 1860 presidential campaign, Lincoln would describe his protest in the Illinois legislature as one that 'briefly defined his position on the slavery question, and so far as it goes, it was then the same that it is now."
Neighbor 2: Lincoln is well known for ending slavery in the United States. In 1861 – 1862, however, he made it clear that the North was fighting the war to preserve the Union, not to abolish slavery.
Neighbor 3: Lincoln, Illinois, is the only city to be named for Abraham Lincoln before he became President. Lincoln's name and image appear in numerous places.
Neighbor 4: Lincoln won the Republican Party nomination in 1860 and was elected president later t

### Running search analysis for query: 'Who was Abraham Adams?' with k=15, rerank=True

Found 15 neighbors.
Neighbor 1: Adams, a sponsor of the American Revolution in Massachusetts, was a driving force for independence in 1776; Jefferson called him the "Colossus of Independence". He represented the Continental Congress in Europe.
Neighbor 2: His birthplace is now part of Adams National Historical Park. His father, a farmer and a Deacon, also named John (1690-1761), was a fourth-generation descendant of Henry Adams, who immigrated from Barton St David, Somerset, England, to Massachusetts Bay Colony in about 1636, from a Welsh male line called Ap Adam.
Neighbor 3: The Works of John Adams (1854), vol III, p 421, diary entry for July 26, 1796. The Unitarian Universalist Historical Society sheds some light on Adams’s religious beliefs.
Neighbor 4: Their children were Abigail (1765-1813); future president John Quincy (1767-1848); Susanna (1768–1770); Charles (1770-1800); Thomas Boylston (1772-1832); and Elizabeth (1775) who was stillborn. Adams was not a popular leader like his

### Running search analysis for query: 'Who was Abraham Adams?' with k=15, rerank=False

Found 15 neighbors.
Neighbor 1: Adams, a sponsor of the American Revolution in Massachusetts, was a driving force for independence in 1776; Jefferson called him the "Colossus of Independence". He represented the Continental Congress in Europe.
Neighbor 2: His birthplace is now part of Adams National Historical Park. His father, a farmer and a Deacon, also named John (1690-1761), was a fourth-generation descendant of Henry Adams, who immigrated from Barton St David, Somerset, England, to Massachusetts Bay Colony in about 1636, from a Welsh male line called Ap Adam.
Neighbor 3: The Works of John Adams (1854), vol III, p 421, diary entry for July 26, 1796. The Unitarian Universalist Historical Society sheds some light on Adams’s religious beliefs.
Neighbor 4: Their children were Abigail (1765-1813); future president John Quincy (1767-1848); Susanna (1768–1770); Charles (1770-1800); Thomas Boylston (1772-1832); and Elizabeth (1775) who was stillborn. Adams was not a popular leader like his

### Running search analysis for query: 'How did Fillmore ascend to the presidency?' with k=1, rerank=True

Found 1 neighbors.
Neighbor 1: Instead, Fillmore joined the anti-immigrant, anti-Catholic American Party, the political organ of the Know-Nothing movement. He ran in the election of 1856 as the party's candidate, attempting to win a non-consecutive second term as President (a feat accomplished only once in American politics, by Grover Cleveland).
Context used for answering 'How did Fillmore ascend to the presidency?':
Context 1: Instead, Fillmore joined the anti-immigrant, anti-Catholic American Party, the political organ of the Know-Nothing movement. He ran in the election of 1856 as the party's candidate, attempting to win a non-consecutive second term as President (a feat accomplished only once in American politics, by Grover Cleveland).
[ChatCompletionChoice(index=0, message=AssistantMessage(content='Fillmore ascended to the presidency upon the death of President Zachary Taylor in 1850.', tool_calls=None, prefix=False, role='assistant'), finish_reason='stop')]
Fillmore ascended to 

### Running search analysis for query: 'How did Fillmore ascend to the presidency?' with k=5, rerank=False

Found 5 neighbors.
Neighbor 1: Instead, Fillmore joined the anti-immigrant, anti-Catholic American Party, the political organ of the Know-Nothing movement. He ran in the election of 1856 as the party's candidate, attempting to win a non-consecutive second term as President (a feat accomplished only once in American politics, by Grover Cleveland).
Neighbor 2: He tried to unite the party by pointing out the differences between the Whigs and the Democrats (by proposing tariff reforms that negatively reflected on the Democratic Party). Another primary objective of Fillmore was to preserve the Union from the intensifying slavery debate.
Neighbor 3: Fillmore signaled this shift by appointing Daniel Webster as his Secretary of State. As president, Fillmore dealt with increasing party divisions within the Whig party; party harmony became one of his primary objectives.
Neighbor 4: Official White House portrait of Millard FillmoreFillmore ascended to the presidency upon the sudden and unexpected

### Running search analysis for query: 'How did Fillmore ascend to the presidency?' with k=10, rerank=True

Found 10 neighbors.
Neighbor 1: Instead, Fillmore joined the anti-immigrant, anti-Catholic American Party, the political organ of the Know-Nothing movement. He ran in the election of 1856 as the party's candidate, attempting to win a non-consecutive second term as President (a feat accomplished only once in American politics, by Grover Cleveland).
Neighbor 2: He tried to unite the party by pointing out the differences between the Whigs and the Democrats (by proposing tariff reforms that negatively reflected on the Democratic Party). Another primary objective of Fillmore was to preserve the Union from the intensifying slavery debate.
Neighbor 3: Fillmore signaled this shift by appointing Daniel Webster as his Secretary of State. As president, Fillmore dealt with increasing party divisions within the Whig party; party harmony became one of his primary objectives.
Neighbor 4: Official White House portrait of Millard FillmoreFillmore ascended to the presidency upon the sudden and unexpecte

### Running search analysis for query: 'How did Fillmore ascend to the presidency?' with k=20, rerank=False

Found 20 neighbors.
Neighbor 1: Instead, Fillmore joined the anti-immigrant, anti-Catholic American Party, the political organ of the Know-Nothing movement. He ran in the election of 1856 as the party's candidate, attempting to win a non-consecutive second term as President (a feat accomplished only once in American politics, by Grover Cleveland).
Neighbor 2: He tried to unite the party by pointing out the differences between the Whigs and the Democrats (by proposing tariff reforms that negatively reflected on the Democratic Party). Another primary objective of Fillmore was to preserve the Union from the intensifying slavery debate.
Neighbor 3: Fillmore signaled this shift by appointing Daniel Webster as his Secretary of State. As president, Fillmore dealt with increasing party divisions within the Whig party; party harmony became one of his primary objectives.
Neighbor 4: Official White House portrait of Millard FillmoreFillmore ascended to the presidency upon the sudden and unexpecte

### Running search analysis for query: 'What trail did Lincoln use a Farmers' Almanac in?' with k=1, rerank=True

Found 1 neighbors.
Neighbor 1: Abraham Lincoln Online Retrieved on March 11, 2007 For several months, Lincoln ran a small store in New Salem, selling tea, coffee, sugar, salt, blue calico, brown muslin, straw hats and whiskey. Beveridge (1928) 1:127-8 Later, he found work as village postmaster and as a surveyor.
Context used for answering 'What trail did Lincoln use a Farmers' Almanac in?':
Context 1: Abraham Lincoln Online Retrieved on March 11, 2007 For several months, Lincoln ran a small store in New Salem, selling tea, coffee, sugar, salt, blue calico, brown muslin, straw hats and whiskey. Beveridge (1928) 1:127-8 Later, he found work as village postmaster and as a surveyor.
[ChatCompletionChoice(index=0, message=AssistantMessage(content='No context', tool_calls=None, prefix=False, role='assistant'), finish_reason='stop')]
No context
----------------------------------------


### Running search analysis for query: 'What trail did Lincoln use a Farmers' Almanac in?' with k=5, rerank=False

Found 5 neighbors.
Neighbor 1: Abraham Lincoln Online Retrieved on March 11, 2007 For several months, Lincoln ran a small store in New Salem, selling tea, coffee, sugar, salt, blue calico, brown muslin, straw hats and whiskey. Beveridge (1928) 1:127-8 Later, he found work as village postmaster and as a surveyor.
Neighbor 2: He had purchased the Sinking Spring Farm in December of 1808 for $200 cash and assumption of a debt. The farm site is now preserved as part of Abraham Lincoln Birthplace National Historic Site.
Neighbor 3: Lincoln believed in the Whig theory of the presidency, which left Congress to write the laws while he signed them, vetoing only those bills that threatened his war powers. Thus, he signed the Homestead Act in 1862, making millions of acres of government-held land in the West available for purchase at very low cost.
Neighbor 4: In the 1920s historical markers were placed at the county lines along the route Lincoln traveled in the eight judicial district. This examp

### Running search analysis for query: 'What trail did Lincoln use a Farmers' Almanac in?' with k=10, rerank=True

Found 10 neighbors.
Neighbor 1: Abraham Lincoln Online Retrieved on March 11, 2007 For several months, Lincoln ran a small store in New Salem, selling tea, coffee, sugar, salt, blue calico, brown muslin, straw hats and whiskey. Beveridge (1928) 1:127-8 Later, he found work as village postmaster and as a surveyor.
Neighbor 2: He had purchased the Sinking Spring Farm in December of 1808 for $200 cash and assumption of a debt. The farm site is now preserved as part of Abraham Lincoln Birthplace National Historic Site.
Neighbor 3: Lincoln believed in the Whig theory of the presidency, which left Congress to write the laws while he signed them, vetoing only those bills that threatened his war powers. Thus, he signed the Homestead Act in 1862, making millions of acres of government-held land in the West available for purchase at very low cost.
Neighbor 4: In the 1920s historical markers were placed at the county lines along the route Lincoln traveled in the eight judicial district. This exam

### Running search analysis for query: 'What trail did Lincoln use a Farmers' Almanac in?' with k=20, rerank=False

Found 20 neighbors.
Neighbor 1: Abraham Lincoln Online Retrieved on March 11, 2007 For several months, Lincoln ran a small store in New Salem, selling tea, coffee, sugar, salt, blue calico, brown muslin, straw hats and whiskey. Beveridge (1928) 1:127-8 Later, he found work as village postmaster and as a surveyor.
Neighbor 2: He had purchased the Sinking Spring Farm in December of 1808 for $200 cash and assumption of a debt. The farm site is now preserved as part of Abraham Lincoln Birthplace National Historic Site.
Neighbor 3: Lincoln believed in the Whig theory of the presidency, which left Congress to write the laws while he signed them, vetoing only those bills that threatened his war powers. Thus, he signed the Homestead Act in 1862, making millions of acres of government-held land in the West available for purchase at very low cost.
Neighbor 4: In the 1920s historical markers were placed at the county lines along the route Lincoln traveled in the eight judicial district. This exam

### Running search analysis for query: 'What is the capital of France?' with k=15, rerank=True

Found 15 neighbors.
Neighbor 1: After his term in the Senate, Monroe was appointed Minister to France from 1794 to 1796. His appointment there was made difficult as he had strong sympathies for the French Revolution, but dutifully maintained President Washington's strict policy of neutrality between Britiain and France.
Neighbor 2: In the late stages of the war he took personal control of negotiations with Germany, especially with the Fourteen Points and the Armistice. He went to Paris in 1919 to create the League of Nations and shape the Treaty of Versailles, with special attention on creating new nations out of defunct empires.
Neighbor 3: It also made several suggestions for specific disputes in Europe on the recommendation of Wilson's foreign policy advisor, Colonel Edward M. House, and his team of 150 advisors known as “The Inquiry.” The points were: Abolition of secret treaties Freedom of the seas Free Trade Disarmament Adjustment of colonial claims (decolonization and national s

### Running search analysis for query: 'What is the capital of France?' with k=15, rerank=False

Found 15 neighbors.
Neighbor 1: After his term in the Senate, Monroe was appointed Minister to France from 1794 to 1796. His appointment there was made difficult as he had strong sympathies for the French Revolution, but dutifully maintained President Washington's strict policy of neutrality between Britiain and France.
Neighbor 2: In the late stages of the war he took personal control of negotiations with Germany, especially with the Fourteen Points and the Armistice. He went to Paris in 1919 to create the League of Nations and shape the Treaty of Versailles, with special attention on creating new nations out of defunct empires.
Neighbor 3: It also made several suggestions for specific disputes in Europe on the recommendation of Wilson's foreign policy advisor, Colonel Edward M. House, and his team of 150 advisors known as “The Inquiry.” The points were: Abolition of secret treaties Freedom of the seas Free Trade Disarmament Adjustment of colonial claims (decolonization and national s

### Analysis

The analysis of the output from the generate_answer() function with different query inputs, k values, and reranking options reveals various insights into the performance of the context-answering pipeline. For the query, “How did Fillmore ascend to the presidency?” the generated answers were consistent across different k values of 1, 5, 10, and 20. All the responses conveyed that Millard Fillmore assumed the presidency following the death of President Zachary Taylor in July 1850. This consistency implies that even a small k value was sufficient to retrieve the appropriate context for this specific question. It also indicates that reranking did not lead to significant changes in the outcome, suggesting that the correct context was consistently retrieved regardless of how many neighbors were used in the initial search.

The impact of reranking became evident when examining queries such as “Who was Abraham Lincoln?” and “Who was Abraham Adams?” In both cases, the answers generated with and without reranking were similar, differing mainly in minor phrasing choices. For example, the difference between mentioning Lincoln’s role in “ending slavery” versus “opposing the expansion of slavery” can be attributed to the reranker’s ability to refine content for nuance. However, in both cases, the core factual information remained unchanged. This suggests that reranking, while beneficial for readability and subtle differences, may not substantially alter the content when the correct context is already present in the retrieved neighbors.

For some of the queries, particularly those seeking factual answers, the generated responses consistently returned “No context.” This occurred in response to questions such as “What trail did Lincoln use a Farmers’ Almanac in?” and “What is the capital of France?” The inability to provide a straightforward answer like “Paris” suggests that the relevant information either does not exist in the current corpus or that the search model struggled to effectively retrieve it. This highlights a gap in either the dataset or the search model’s ability to locate concise, factual information, particularly for direct questions.

Another notable observation concerns the query about Abraham Adams. The responses indicated that "Abraham Adams is not mentioned," which is actually a positive aspect of the model's response. The model correctly identifies that this entity does not exist within the corpus, showing that it does not fabricate information when it lacks the necessary context. This accuracy is crucial in maintaining the reliability of generated answers. This reliance on corpus content also means that if a specific entity or topic is not present in the corpus, the model will truthfully indicate its absence, rather than generating a misleading or incorrect response.

Regarding the question, "What is the capital of France?" the model's response of "No context" is due to the hardcoded behavior within the generate_answer function when no relevant context is found. This does not reflect the model's inability to provide factual information like "Paris," but rather the logic imposed by the current implementation of the code. The function is designed to output "No context" whenever it cannot find relevant information within the indexed data, even if the model itself has the knowledge to answer such straightforward questions, to minimize hallucinations.

In conclusion, the performance of the generate_answer() function demonstrates both strengths and areas for improvement. While the model performs well for queries where context is available, it struggles with providing concise factual answers and retrieving relevant information for uncommon entities. Addressing these limitations through improved corpus coverage, external fact integration, and reranking enhancements can help create a more robust and effective context-answering pipeline.